packages/runtime/container-runtime/src/scheduleManager.ts

```diff
@@ -0,0 +1,310 @@
+import EventEmitter from "events";
+import { ITelemetryLogger } from "@fluidframework/common-definitions";
+import { assert, IsoBuffer, performance } from "@fluidframework/common-utils";
+import { IDeltaManager } from "@fluidframework/container-definitions";
+import { DataCorruptionError, extractSafePropertiesFromMessage } from "@fluidframework/container-utils";
+import { isUnpackedRuntimeMessage } from "@fluidframework/driver-utils";
+import { ISequencedDocumentMessage, IDocumentMessage } from "@fluidframework/protocol-definitions";
+import { ChildLogger } from "@fluidframework/telemetry-utils";
+import { compress } from "lz4js";
+import { latencyThreshold } from "./connectionTelemetry";
+import { DeltaScheduler } from "./deltaScheduler";
+import { pkgVersion } from "./packageVersion";
+
+type IRuntimeMessageMetadata = undefined | {
+    batch?: boolean;
+};
+
+/**
+ * This class controls pausing and resuming of inbound queue to ensure that we never
+ * start processing ops in a batch IF we do not have all ops in the batch.
+ */
+ class ScheduleManagerCore {
+    private pauseSequenceNumber: number | undefined;
+    private currentBatchClientId: string | undefined;
+    private localPaused = false;
+    private timePaused = 0;
+    private batchCount = 0;
+
+    constructor(
+        private readonly deltaManager: IDeltaManager<ISequencedDocumentMessage, IDocumentMessage>,
+        private readonly logger: ITelemetryLogger,
+    ) {
+        // Listen for delta manager sends and add batch metadata to messages
+        this.deltaManager.on("prepareSend", (messages: IDocumentMessage[]) => {
+            if (messages.length === 0) {
+                return;
+            }
+
+            // First message will have the batch flag set to true if doing a batched send
+            const firstMessageMetadata = messages[0].metadata as IRuntimeMessageMetadata;
+            if (!firstMessageMetadata?.batch) {
+                return;
+            }
+
+            // If the batch contains only a single op, clear the batch flag.
+            if (messages.length === 1) {
+                delete firstMessageMetadata.batch;
```

Comment:
One interesting observation: in general, we want to compress all ops that are coming from runtime.
You can't do it here unconditionally, because that will capture such things as Summarize and Noop ops, and their content should not be compressed (at least it will not work without service supporting that).

So my preference would be not to do compression here but do compression where we send ops (ContainerRuntime.submitRuntimeMessage). That way you will compress individual ops.

As for shifting content of all ops in a batch - let's handle it separately. It would be possible to do so in ContainerRuntime.submitRuntimeMessage, though it would be a bit ugly.